static/resources/harness.js

```diff
@@ -84,182 +102,74 @@
-              } else {
-                thisCompiled = '"'+property+'" in '+parentPrefix+subtest.scope;
-                value = eval(thisCompiled);
-              }
-
-              if (!compiled) {
-                // Set to first compiled statement in case support is false
-                compiled = thisCompiled;
-              }
-            }
-
-            result.result = value;
-            if (value === true) {
-              compiled = thisCompiled;
-
-              if (subtest.scope === 'CSS.supports') {
-                if (prefix) {
-                  parentPrefix = '-' + prefix + '-';
-                } else {
-                  parentPrefix = '';
-                }
-              } else {
-                parentPrefix = prefix;
-              }
-              break;
-            }
-          }
-
-          compiledCode.push(compiled);
+          result.result = null;
+          result.message = 'returned ' + stringify(value);
         }
+      } catch (err) {
+        result.result = null;
+        result.message = 'threw ' + stringify(err);
+      }
 
-        if (result.result === false) {
-          break;
-          // Tests are written in hierarchy order, so if the parent (first
-          // test) is unsupported, so is the child (next test)
+      if (result.result !== false) {
+        result.info.code = test.code;
+        if (test.prefix) {
+          result.info.prefix = test.prefix;
         }
-
-        result.prefix = parentPrefix;
+        break;
       }
-    } catch (err) {
-      result.result = null;
-      result.message = 'threw ' + stringify(err);
     }
 
     if (data.info !== undefined) {
       result.info = Object.assign({}, result.info, data.info);
     }
 
-    result.info.code = compiledCode.join(' && ');
+    if (result.result === false) {
+      result.info.code = data.tests[0].code;
+    }
     result.info.scope = data.scope;
 
     return result;
   }
 
-  function runCSS(callback) {
-    var results = [];
-
+  function runCSS(callback, results) {
     var length = pending.length;
     for (var i = 0; i < length; i++) {
-      updateStatus('Testing ' + pending[i].name);
-      results.push(test(pending[i]));
+      if (pending[i].scope == 'CSS') {
+        updateStatus('Testing ' + pending[i].name);
+        results.push(test(pending[i]));
+      }
     }
 
-    pending = [];
-
     callback(results);
   }
 
-  function runWindow(callback) {
-    var results = [];
-
+  function runWindow(callback, results) {
     var length = pending.length;
     for (var i = 0; i < length; i++) {
-      updateStatus('Testing ' + pending[i].name);
-      results.push(test(pending[i]));
+      if (pending[i].scope == 'Window') {
+        updateStatus('Testing ' + pending[i].name);
+        results.push(test(pending[i]));
+      }
     }
 
-    pending = [];
-
     callback(results);
   }
 
-  function runWorker(callback) {
-    var results = [];
+  function runWorker(callback, results) {
```

Comment:
Not in this PR, but the setup for worker tests I'm imaging when most of the of work is done on the server side is that a single large script is run in a worker, and then the results are all posted back in one go. In the window context there wouldn't be any need to keep track of pending tests.